### PR TITLE
hubble: add trace reason support in hubble flows

### LINKED ISSUE #31202: Hubble: expose the trace reason field
https://github.com/cilium/cilium/pull/30154 and https://github.com/cilium/cilium/pull/31073 introduced new trace reasons, before that all trace reasons were CT related and used to infer some Hubble specific fields like `traffic_direction` and `is_reply`. `cilium monitor` already show the trace reason in its output, and adding it to hubble would close the gap between the two.

Succes criteria:

- [ ] New trace reason field on `flowpb.Flow`

### PR DESCRIPTION
**Pull request by kaworu**
<del>_**Reviewer note**: this PR is a draft as the patch are on top of https://github.com/cilium/cilium/pull/31211, and I wish to wait for https://github.com/cilium/cilium/pull/31073 to be merged and handle conflicts in this PR._</del>

See commits, closes https://github.com/cilium/cilium/issues/31202.

### CODE REVIEW

**Review by bimmlerd — COMMENTED** [review]
CODEOWNERS lgtm

**Review by bimmlerd — APPROVED** [review]

**Review by rolinh — APPROVED** [review]
Looks solid

File: api/v1/flow/flow.proto [hunk 1]
```diff
@@ -183,6 +185,18 @@ enum TraceObservationPoint {
     TO_NETWORK = 11;
 }
 
+enum TraceReason {
```
Comment (ghost) [inline api/v1/flow/flow.proto:193]: Is there a reason behind the ordering here? Intuitively I'd expect `TRACE_REASON_UNKNOWN` to be `0` (or maybe `TRACE_REASON_UNSPECIFIED` to follow [proto best practices](https://protobuf.dev/programming-guides/dos-donts/#unspecified-enum)).

**Review by ghost — COMMENTED** [review]
Looks good, left one question

File: api/v1/flow/flow.proto (hunk 1 above)
Comment (kaworu) [reply api/v1/flow/flow.proto:193]: Ordering / values are mapped 1:1 with the datapath, see [here](https://github.com/cilium/cilium/blob/9e89397d2d81fe915bfbd74d41bd72e5d0c6ad5b/pkg/monitor/datapath_trace.go#L136-L146) and [here](https://github.com/cilium/cilium/blob/9e89397d2d81fe915bfbd74d41bd72e5d0c6ad5b/bpf/lib/trace.h#L47-L62). I thought we were doing this consistently (e.g. drop reason), but after a second look it's not the case for trace observation point, so maybe it's ok to shuffle the order a bit.

**Comment by kaworu** [comment]
@lambdanis addressed your feedback and now setting `TRACE_REASON_UNKNOWN = 0` at the protobuf level. The decoding logic is consequently a tad more complex, please take another look :pray:

**Review by ghost — APPROVED** [review]
Makes sense, thanks @kaworu!

I see there is a conflict, but apart from that looks good to me.

File: pkg/hubble/parser/threefour/parser.go [hunk 2]
```diff
@@ -435,6 +432,29 @@ func decodeCiliumEventType(eventType, eventSubType uint8) *pb.CiliumEventType {
...
+	// values before monitor.TraceReasonUnknown are "offset by one", e.g.
+	// TraceReasonCtEstablished = 1 → TraceReason_ESTABLISHED = 2 to make room
+	// the the zero value.
+	case tn.TraceReason() < monitor.TraceReasonUnknown:
+		return pb.TraceReason(tn.TraceReason()) + 1
+		// all values greater than monitor.TraceReasonUnknown are mapped 1:1
+		// with the datapath values.
+	default:
+		return pb.TraceReason(tn.TraceReason())
+	}
```
Comment (glibsm) [inline pkg/hubble/parser/threefour/parser.go]: woof. Makes sense after I actually checked out the code and looked at values, sorry you had to deal with this.

**Review by glibsm — APPROVED** [review]

**Review by gentoo-root — APPROVED** [review]

**Pull request merged**